### PR TITLE
Configurable SSH cipher suite

### PR DESCRIPTION
**Pull request by spacetourist**
As requested, merging change from Gogs.io

### CODE REVIEW

**Comment by lunny** [comment]
LGTM

File: conf/app.ini [hunk 1]
```diff
@@ -110,6 +110,8 @@ SSH_PORT = 22
 SSH_LISTEN_PORT = %(SSH_PORT)s
 ; Root path of SSH directory, default is '~/.ssh', but you have to use '/home/git/.ssh'.
 SSH_ROOT_PATH =
+; Choose the ciphers to support for SSH connections
```
Comment (bkcsoft) [inline conf/app.ini]: Please make it clear that this only affects Built-In SSH-server

File: routers/init.go [hunk 2]
```diff
@@ -75,7 +75,7 @@ func GlobalInit() {
 	checkRunMode()
 
 	if setting.InstallLock && setting.SSH.StartBuiltinServer {
-		ssh.Listen(setting.SSH.ListenHost, setting.SSH.ListenPort)
-		log.Info("SSH server started on %s:%v", setting.SSH.ListenHost, setting.SSH.ListenPort)
+               ssh.Listen(setting.SSH.ListenHost, setting.SSH.ListenPort, setting.SSH.ServerCiphers)
+               log.Info("SSH server started on %s:%v. Cipher list (%v)", setting.SSH.ListenHost, setting.SSH.ListenPort, setting.SSH.ServerCiphers)
```
Comment (bkcsoft) [inline routers/init.go]: ListenPort should use `%d` not `%v`

**Review by bkcsoft — CHANGES_REQUESTED** [review]

**Comment by bkcsoft** [comment]
Not exactly happy about having configs that only affect the build-in server, but for this it's hard to do otherwise 🙁

File: routers/init.go [hunk 3]
```diff
@@ -75,7 +75,7 @@ func GlobalInit() {
 	checkRunMode()
 
 	if setting.InstallLock && setting.SSH.StartBuiltinServer {
-		ssh.Listen(setting.SSH.ListenHost, setting.SSH.ListenPort)
-		log.Info("SSH server started on %s:%v", setting.SSH.ListenHost, setting.SSH.ListenPort)
+               ssh.Listen(setting.SSH.ListenHost, setting.SSH.ListenPort, setting.SSH.ServerCiphers)
```
Comment (ethantkoenig) [inline routers/init.go]: Please use tabs for indentation

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

**Comment by aaronjwood** [comment]
Can I suggest that you prioritize AES-GCM above everything else? I think an authenticated encryption mode should be first on everyone's list.

It might be nice to prevent the use of the arcfour suites entirely. Since most modern chips have the AES-NI instruction set to accelerate AES operations most people would see better performance too.

**Comment by lunny** [comment]
@aaronjwood Yes. I think the default values should be security enough.

**Comment by lunny** [comment]
It seems no response here, let's move it to v1.2

**Comment by evantbyrne** [comment]
According to the Go SSH documentation on [ssh.Config](https://godoc.org/golang.org/x/crypto/ssh#Config),

```
// The allowed cipher algorithms. If unspecified then a sensible
// default is used.
Ciphers []string
```

So it might be best to leave that ini option undefined _by default_, unless this cipher set differs from the default Go one and there is a specific rationale for such.

File: modules/setting/setting.go [hunk 4]
```diff
@@ -618,6 +619,7 @@ please consider changing to GITEA_CUSTOM`)
 	}
 
 	SSH.RootPath = path.Join(homeDir, ".ssh")
+	SSH.ServerCiphers = sec.Key("SSH_SERVER_CIPHERS").Strings(",")
```
Comment (tboerger) [inline modules/setting/setting.go]: Requires a default value?

File: conf/app.ini (hunk 1 above)
Comment (cdslashetc) [reply conf/app.ini]:
```
; For built-in SSH server only, choose the ciphers to support for SSH connections,
; for system SSH this setting has no effect
```

**Comment by strk** [comment]
It's somewhat ironic that Gogs is ahead of the community fork made to speed things up...
How about reviewers are less picky unless they have time to fix things themselves ?
I'll start with my LGTM here

**Comment by lunny** [comment]
As @evantbyrne said, keep the default as Go did maybe a good choice. @spacetourist maybe you can change this?

**Comment by sapk** [comment]
Any news ?

**Comment by lunny** [comment]
If @evantbyrne no response. I think maintainers could do something.
I think this PR needs two fixes:
1. add more comment as @cdslashetc said.
2. change the default value as @evantbyrne said

**Comment by sapk** [comment]
+ fix indentation

**Comment by lafriks** [comment]
@bkcsoft @ethantkoenig fixed

**Review by ethantkoenig — APPROVED** [review]

**Review by bkcsoft — APPROVED** [review]

**Comment by lafriks** [comment]
Make LG-TM work again

**Comment by ncwgf** [comment]
This patch will report below error using SSH to `git fetch` who is not from fresh install
because `conf/app.ini` missed config `SSH_SERVER_CIPHERS`
problem solved by manually copy the setting from this commit to `conf/app.ini`
```
git fetch gitea
FATAL ERROR: Couldn't agree a client-to-server cipher (available: )
fatal: Could not read from remote repository.
```

**Pull request merged**